server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/Dequeuer.java

```diff
@@ -90,31 +90,29 @@ public void done(boolean success) {
     }
 
     Flux<? extends MailQueue.MailQueueItem> deQueue() {
-        return flux.concatMap(this::loadItem)
-            .concatMap(this::filterIfDeleted);
+        return flux.flatMapSequential(this::loadItem)
+            .flatMapSequential(this::itemWithIsPresent)
+            .concatMap(pair -> keepWhenPresent(pair.getValue(), pair.getKey()));
```

Comment:
I guess you can merge `itemWithIsPresent()` and `keepWhenPresent()` into one method for the sake of simplicity and it can avoid Pair intermediate representation.

Reply:
nope, because they don't have the same sequential requirements.